### PR TITLE
Point testing dependencies to `main`

### PR DESCRIPTION
**Pull request by mikealfare**
All testing dependencies should point to main to ensure that the latest changes are reflected.

- Move all optional dependencies into hatch environments
  - They aren't production dependencies
  - We can't install from GitHub in optional dependencies and also publish to PyPI
- Make the `default` environment the dev version of all production dependencies
- Inherit from the `default` environment for non-detached hatch environments (typechecking and testing)
- Pin lower bound for dependencies in alignment with our practices

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-postgres/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] This PR has no interface changes (e.g. macros, cli, logs, json artifacts, config files, adapter interface, etc) or this PR has already received feedback and approval from Product or DX

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -87,44 +50,68 @@ packages = ["dbt"]
...
 
 [tool.hatch.envs.default]
-features = [
-    "lint",
-    "typecheck",
-    "test",
-    "integration",
-    "build",
+dependencies = [
+    "dbt-adapters @ git+https://github.com/dbt-labs/dbt-adapters.git@tests/get-catalog",
```
Comment (mikealfare) [inline pyproject.toml:54]: Point this to `main` before merging.

File: pyproject.toml [hunk 2]
```diff
@@ -87,44 +50,68 @@ packages = ["dbt"]
...
+]
 [tool.hatch.envs.unit-tests.scripts]
 all = "python -m pytest {args:tests/unit}"
 
 [tool.hatch.envs.integration-tests]
-# TODO: confirm this works for production testing or add appropriate hatch envs
-features = ["dev", "test", "integration"]
+template = "unit-tests"
+extra-dependencies = [
+    "dbt-tests-adapter @ git+https://github.com/dbt-labs/dbt-adapters.git@tests/get-catalog#subdirectory=dbt-tests-adapter",
```
Comment (mikealfare) [inline pyproject.toml:98]: Point this to `main` before merging.

File: tests/unit/test_adapter.py
Comment (mikealfare) [inline tests/unit/test_adapter.py:1]: The test cases added here are being migrated back to `dbt-postgres` (for now). It was determined that the issue was likely being caused by a combination of not looking at `main` for internal dependencies and not instantiating the invocation context variables before checking them. `set_invocation_context({})` is run on line 338 now, right before getting the catalog, and tests are now passing.

File: pyproject.toml
Comment (mikealfare) [inline pyproject.toml:1]: Dependencies were moved down into the hatch environments, in alignment with `dbt-adapters`. This was done for a combination of reasons:
- we cannot publish a package to PyPI with dependencies installed from GitHub, so development versions of internal packages would need to be identified as hatch environment dependencies
- there is not a lot of value in installing dev and test tooling without the development versions of internal packages
- having just the tooling as installable extras in the production distribution is confusing

File: pyproject.toml [hunk 3]
```diff
@@ -87,44 +50,68 @@ packages = ["dbt"]
...
+]
 [tool.hatch.envs.typecheck.scripts]
 all = "python -m mypy ."
 
 [tool.hatch.envs.unit-tests]
-# TODO: confirm this works for production testing or add appropriate hatch envs
-features = ["dev", "test"]
+dependencies = [
+    # TODO: remove `dbt-core` dependencies from unit tests
+    "dbt-core @ git+https://github.com/dbt-labs/dbt-core.git#subdirectory=core",
```
Comment (colin-rogers-dbt) [inline pyproject.toml:85]: kinda surprised we don't need dbt-tests-adapter for unit tests

**Review by colin-rogers-dbt — APPROVED** [review]

**Pull request merged**